src/stack/array_stack.rs

```diff
@@ -0,0 +1,134 @@
+use crate::stack::Stack;
+
+/// Uses a Vector of a given capacity (16 by default)
+struct ArrayStack<T: Clone> {
+    array: Vec<Option<T>>,
+    head: Option<usize>,
+    size: usize,
+    capacity: usize,
+}
+
+impl<T: Clone> Stack<T> for ArrayStack<T> {
+    /// Pushes item of type `T` at the position pointed by `head`
+    /// Capacity will be updated once `capacity == count`
+    /// Amortized complexity: O(1)
+    fn push(&mut self, item: T) -> () {
+        if self.capacity() == self.count() {
+            self.update_capacity(self.capacity * 2);
```

Comment:
Here `self.capacity` is used, instead of `self.capacity()`, to read the capacity value, while the method is used, conversely, at the line before. 
Unless there is a specific reason not to use the method, I would consistently go through it in this context, since `push` doesn't seem to need to know about the backing field of the capacity property.
Alternatively, the backing field could be used directly and accessor methods removed, depending on what is more idiomatic in Rust.